mythril/laser/smt/__init__.py

```diff
@@ -65,24 +75,30 @@ class _SmtSymbolFactory(SymbolFactory):
     """
 
     @staticmethod
-    def Bool(value: bool, annotations=None):
+    def Bool(
+        value: "__builtins__.bool", annotations: Annotations = None
+    ) -> Union[bool.Bool, z3.BoolRef]:
         """
         Creates a Bool with concrete value
         :param value: The boolean value
         :param annotations: The annotations to initialize the bool with
         :return: The freshly created Bool()
         """
-        raw = z3.Bool(value)
+        raw = z3.BoolVal(value)
         return Bool(raw, annotations)
 
     @staticmethod
-    def BitVecVal(value: int, size: int, annotations=None):
+    def BitVecVal(
+        value: int, size: int, annotations: Annotations = None
+    ) -> Union[BitVec, z3.BitVecRef]:
         """Creates a new bit vector with a concrete value."""
         raw = z3.BitVecVal(value, size)
         return BitVec(raw, annotations)
 
     @staticmethod
-    def BitVecSym(name: str, size: int, annotations=None):
+    def BitVecSym(
+        name: str, size: int, annotations: Annotations = None
+    ) -> Union[BitVec, z3.BitVecRef]:
```

Comment:
this doesn't return z3.BitVecRef right?

Reply:
The BitVecSym function is defined in the parent SymbolFactory class. In the _SmtSymbolFactory subclass BitVecSym returns a BitVec and in the _Z3SymbolFactory subclass BitVecSym returns a z3.BitVecRef.

However, the signature of BitVecSym in the base class must be compatible with all of the implementing classes, ergo why I chose to make it a union.

However now that you point it out, we CAN be more precise here! We can do the following:

```
T=TypeVar("T", bound=Union[BitVec, z3.BitVecRef])
class SymbolFactory(Generic[T]):
    def BitVecSym(..)->T:
...
class _SmtSymbolFactory(SymbolFactory[BitVec]):
    ...
class _Z3SymbolFactory(SymbolFactory[z3.BitVecRef]):
    ...
```

That may make this simpler. Do you want me to try that instead?

Reply:
yes, that would be great.